### PR TITLE
feat(new-rule): ibm-api-symmetry

### PR DESCRIPTION
**Pull request by dpopp07**
Ensures that Summary, Prototype, and Patch schemas are proper graph fragments of the canonical schema they are variants of.

### General checklist
Please make sure that your PR fulfills the following requirements:  
- [x] The commit message follows the [Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Dependencies have been updated as needed
- [x] .secrets.baseline updated as needed?

#### Checklist for adding a new validation rule:
- [x] Added new validation rule definition (packages/ruleset/src/rules/*.js, index.js)
- [x] If necessary, added new validation rule implementation (packages/ruleset/src/functions/*.js, updated index.js)
- [x] Added new rule to default configuration (packages/ruleset/src/ibm-oas.js)
- [x] Added tests for new rule (packages/ruleset/test/*.test.js)
- [x] Added docs for new rule (docs/ibm-cloud-rules.md)

### CODE REVIEW

File: packages/ruleset/src/functions/api-symmetry.js [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+  // The function is wrapped in a partial so that we can track `variantOmitsProperties`
+  // across the entire schema without overriding it during recursive invocations of the
+  // graph fragment checker.
+
+  function isGraphFragment(variant, canonical, canonicalPath, fromApplicator) {
+    // A variant schema cannot allow extraneous properties and still be
+    // an explicit graph fragment of the canonical schema.
+    if (variant.additionalProperties) {
+      logger.info(
```
Comment (dpopp07) [inline packages/ruleset/src/functions/api-symmetry.js:158]: I use `info` level messages here to be able to tell the user all the reasons why the schema isn't a graph fragment without trying to a) include all the reasons in the primary message or b) making the user sift through the debug logs to see the reasons. I'm including a note in the docs that info level logs will give the user more specific info for this rule

File: packages/ruleset/src/functions/schema-naming-convention.js [hunk 2]
```diff
@@ -298,143 +290,6 @@ function checkSchemaNames(apidef, nodes) {
   return errors;
 }
 
-function collectResourceOrientedPaths(apidef) {
```
Comment (dpopp07) [inline packages/ruleset/src/functions/schema-naming-convention.js:301]: All of these functions were moved to the `utils` folder for sharing between functions

**Review by padamstx — APPROVED** [review]
Looks great!
Despite my best efforts (or at least slighly more than mediocre efforts :joy:) I couldn't find anything wrong :smile:

File: docs/ibm-cloud-rules.md [hunk 3]
```diff
@@ -743,6 +750,113 @@ This section provides reference documentation about the IBM Cloud Validation Rul
...
+<td><b>ibm-api-symmetry</b></td>
+</tr>
+<tr>
+<td valign=top><b>Description:</b></td>
+<td>
+Each resource-representing schema should follow the <a href="https://cloud.ibm.com/docs/api-handbook?topic=api-handbook-schemas">IBM Cloud API Handbook schema structure conventions</a>. Specifically, this rule ensures that Summary, Prototype, and Patch schemas are proper graph fragments of the canonical schema they are variants of. According to the Handbook, "a graph fragment schema has the same structure as its canonical schema with some properties omitted from the schema or from any nested object schemas."
+
+This convention ensures that resource-oriented APIs are symmetrical between requests and responses in how resources are represented.
+
+Note: the rule will report each violation of the graph fragment pattern through "info" level logs. Pass the option <code>--log-level ibm-api-symmetry=info</code> to see more info about any violations that occur.
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply docs/ibm-cloud-rules.md:766]: Ah good catch! That's a good idea, I will make that change

**Pull request merged**